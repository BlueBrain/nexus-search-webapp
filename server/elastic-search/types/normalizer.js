import palettes from "distinct-colors";

const COLOR_SETTINGS = {
  chromaMax: 80,
  lightMin: 70
};

/**
 *
 *
 * @param {Object} docs elastic search response
 * @returns {Array} types
 */
function normalizer (docs) {
  let types = docs.aggregations["@types"].buckets;
  // TODO move to client?
  let palette = palettes({
    count: types.length,
    ...COLOR_SETTINGS
  });
  types = types.map((type, index) => {
    type.color = palette[index].hex();
    return type;
  });
  return types
}

export default normalizer;