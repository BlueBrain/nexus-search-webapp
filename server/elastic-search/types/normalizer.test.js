import normalizer from "./normalizer";

const MOCK_ES_RESPONSE = {
  aggregations: {
    "@types": {
      doc_count_error_upper_bound: 0,
      sum_other_doc_count: 0,
      buckets: [
        {
          key: "prov:Entity",
          doc_count: 30
        },
        {
          key: "nsg:ResponseTrace",
          doc_count: 25
        },
        {
          key: "nsg:Trace",
          doc_count: 25
        },
        {
          key: "nsg:Subject",
          doc_count: 3
        },
        {
          key: "nsg:ReconstructedWholeBrainCell",
          doc_count: 2
        },
        {
          key: "prov:Agent",
          doc_count: 1
        },
        {
          key: "schema:Dataset",
          doc_count: 1
        }
      ]
    }
  }
};

describe("normalizer()", () => {
  it("returns a list of buckets", () => {
    expect(Array.isArray(normalizer(MOCK_ES_RESPONSE))).toBe(true);
  });

  it("matches the expected output with colors added", () => {
    expect(normalizer(MOCK_ES_RESPONSE)).toMatchSnapshot();
  });
});
