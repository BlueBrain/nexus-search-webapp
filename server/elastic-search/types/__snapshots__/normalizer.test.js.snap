// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizer() matches the expected output with colors added 1`] = `
Array [
  Object {
    "color": "#dcdadd",
    "doc_count": 30,
    "key": "prov:Entity",
  },
  Object {
    "color": "#ded76b",
    "doc_count": 25,
    "key": "nsg:ResponseTrace",
  },
  Object {
    "color": "#f6b0e7",
    "doc_count": 25,
    "key": "nsg:Trace",
  },
  Object {
    "color": "#78e6a8",
    "doc_count": 3,
    "key": "nsg:Subject",
  },
  Object {
    "color": "#f4be9c",
    "doc_count": 2,
    "key": "nsg:ReconstructedWholeBrainCell",
  },
  Object {
    "color": "#5bebee",
    "doc_count": 1,
    "key": "prov:Agent",
  },
  Object {
    "color": "#78ccf6",
    "doc_count": 1,
    "key": "schema:Dataset",
  },
]
`;
